File: csharp/GeoSample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MongoDB.Bson;
using MongoDB.Driver;
using System.Security.Authentication;

namespace TestMongoApp
{
    class GeoSample
    {
        private const string DBName = "testdb";
        private const string CollName = "csharpsample";
        private MongoClient client;
        private IMongoDatabase db;
        private IMongoCollection<SampleDocument> coll;

        public GeoSample()
        {
            
        }
        public async Task Run()
        {
            try
            {
                this.client = new MongoClient(this.GetSettingsHelper());
                this.db = this.client.GetDatabase(GeoSample.DBName);
                this.coll = this.db.GetCollection<SampleDocument>(GeoSample.CollName);

                
                await this.PopulateInitialDocs();
                    
                await this.ReadUsingReadPref(ReadPreferenceMode.Primary);
                await this.ReadUsingReadPref(ReadPreferenceMode.Nearest);
                await this.ReadUsingReadPref(ReadPreferenceMode.Secondary);
                    
                await this.ReadFromSpecificRegion("West US");

                await this.Cleanup();

                Console.WriteLine("Run complete");
                Console.ReadLine();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.GetType());
            }
        }

        private MongoClientSettings GetSettingsHelper()
        {
            const string MongoAuthScramSHATypeString = "SCRAM-SHA-1";
            List<MongoServerAddress> servers = new List<MongoServerAddress>();

            servers.Add(new MongoServerAddress("mongogeotest.documents-bvt.windows-int.net", 10255));
            MongoClientSettings settings = new MongoClientSettings();
            settings.Servers = servers;

            settings.ConnectionMode = ConnectionMode.ReplicaSet;
            settings.ReplicaSetName = "globaldb";

            settings.UseSsl = true;
            settings.VerifySslCertificate = false;
            settings.SslSettings = new SslSettings();
            settings.SslSettings.EnabledSslProtocols = SslProtocols.Tls12;

            MongoIdentity identity1 = new MongoInternalIdentity(GeoSample.DBName, "mongogeotest ");
            // [SuppressMessage("Microsoft.Security", "CS002:SecretInNextLine")]
            MongoIdentityEvidence evidence = new PasswordEvidence("jU4DlJ1Fwy7qcovsJcNi9CD6CSkdh7bB00Z8Xm8drBaiNnEmuNK8hhrBna90ESIfWK0NiPtStCRzn5iuhm9Vkw==");

            settings.Credentials = new List<MongoCredential>()
            {
                new MongoCredential(MongoAuthScramSHATypeString, identity1, evidence),
            };

            return settings;
        }

        private async Task PopulateInitialDocs()
        {
            for (int i = 0; i < 10; i++)
            {
                await this.coll.InsertOneAsync(SampleDocument.getADoc());
            }

            Console.WriteLine("Insert completed");
        }

        private async Task<int> ReadUsingReadPref(ReadPreferenceMode readPrefMode)
        {
            Console.WriteLine("ReadPrefMode: {0}", readPrefMode.ToString());
            var filter = new BsonDocument();
            int count = 0;
            ReadPreference readpref = new ReadPreference(readPrefMode);
            using (var cursor = await this.coll.WithReadPreference(readpref).FindAsync(filter))
            {
                while (await cursor.MoveNextAsync())
                {
                    var batch = cursor.Current;
                    foreach (var doc in batch)
                    {
                        count++;
                        Console.WriteLine(doc);
                    }
                }
            }
            return count;
        }

        private async Task ReadFromSpecificRegion(string regionName)
        {

            Console.WriteLine("Read from region: {0}", regionName);
            var filter = new BsonDocument();
            int count = 0;
            ReadPreference readpref = ReadPreference.Nearest.With(GeoSample.GetRegionTagSets(regionName));
            using (var cursor = await this.coll.WithReadPreference(readpref).FindAsync(filter))
            {
                while (await cursor.MoveNextAsync())
                {
                    var batch = cursor.Current;
                    foreach (var doc in batch)
                    {
                        Console.WriteLine(doc);
                        count++;
                    }
                }
            }
        }

        private static List<TagSet> GetRegionTagSets(string regionName)
        {
            Tag regionTag = new Tag("region", regionName);
            List<Tag> tags = new List<Tag>();
            tags.Add(regionTag);

            TagSet tagset = new TagSet(tags);
            List<TagSet> tagsets = new List<TagSet>();
            tagsets.Add(tagset);

            return tagsets;
        }

        private async Task Cleanup()
        {
            var dbuilder = Builders<SampleDocument>.Filter;
            var delete = dbuilder.Empty;
            var deleteResult = await this.coll.DeleteManyAsync(delete);
        }


        private sealed class SampleDocument
        {
            public BsonObjectId _id { get; set; }
            public string payload { get; set; }

            public override bool Equals(object obj)
            {
                if (obj == null)
                    return false;

                SampleDocument other = obj as SampleDocument;
                if (other == null)
                    return false;

                if (other._id == this._id && other.payload.Equals(this.payload))
                    return true;
                else
                    return false;
            }
            public override int GetHashCode()
            {
                return _id.GetHashCode() ^ payload.GetHashCode();
            }

            public static SampleDocument getADoc()
            {
                SampleDocument sd = new SampleDocument();
                sd.payload = new String(Enumerable.Repeat('X', 10).ToArray());
                return sd;
            }

            public static string getAPayload()
            {
                return new String(Enumerable.Repeat('X', 10).ToArray());
            }

            public override string ToString()
            {
                return string.Format("_id: {0} ; payload: {1}", this._id, this.payload);
            }
        }
    }
}

File: java/GeoSample.java
package mongogeotest;

import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import org.bson.Document;

import com.mongodb.MongoClient;
import com.mongodb.MongoClientURI;
import com.mongodb.ReadPreference;
import com.mongodb.Tag;
import com.mongodb.TagSet;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;


public class GeoSample {
	
   public static void main(String[] args)
   {
	   try 
	   {	   
		   
		   MongoClientURI uri = new MongoClientURI("mongodb://<username>:<password>@<DocDBAccountEndpoint>:<port>/<dbname>?ssl=true&replicaSet=globaldb");
		   MongoClient mongoClient = new MongoClient(uri);
		   
		   MongoDatabase database = mongoClient.getDatabase("<dbname>");
		   
		   //Example of inserting docs (goes to write region)
		   MongoCollection<Document> collection = database.getCollection("<collname>");
		   List<Document> docs = new ArrayList<Document>();
		   docs.add(new Document("fruit", "apple"));
		   docs.add(new Document("fruit", "orange"));
		   docs.add(new Document("fruit", "mango"));
		   collection.insertMany(docs);
		   
		   //Example of reading from Write region (readpreference = primary)
		   MongoCollection<Document> writeRegionCollection = database.getCollection("<collname>").withReadPreference(ReadPreference.primary());
		   List<Document> foundDocument = writeRegionCollection.find().into(new ArrayList<Document>());
		   System.out.println("Documents from write region:");
		   for(Iterator<Document> idoc = foundDocument.iterator(); idoc.hasNext();)
		   {
			    Document item = idoc.next();
			    System.out.println(item);
		   }
		   
		   //Example of reading from Nearest region (readpreference = nearest)
		   MongoCollection<Document> nearestRegionCollection = database.getCollection("<collname>").withReadPreference(ReadPreference.nearest());
		   foundDocument = nearestRegionCollection.find().into(new ArrayList<Document>());
		   System.out.println("Documents from nearest region:");
		   for(Iterator<Document> idoc = foundDocument.iterator(); idoc.hasNext();)
		   {
			    Document item = idoc.next();
			    System.out.println(item);
		   }	   

		   
		   //Example of reading from Read region (readpreference = secondary)
		   MongoCollection<Document> readRegionCollection = database.getCollection("<collname>").withReadPreference(ReadPreference.secondary());
		   foundDocument = readRegionCollection.find().into(new ArrayList<Document>());
		   System.out.println("Documents from read region:");
		   for(Iterator<Document> idoc = foundDocument.iterator(); idoc.hasNext();)
		   {
			    Document item = idoc.next();
			    System.out.println(item);
		   }
		   
		   //Example of reading from a specific region using tags
		   //you specify your read region here as a Tag.
		   //pick region name based on your DocumentDB region from:
		   //https://azure.microsoft.com/en-us/regions/
		   //or use portal for region names
		   Tag tag = new Tag("region", "<regionName>"); 
		   TagSet tagset = new TagSet(tag);
		   List<TagSet> tagsetList = new ArrayList<TagSet>();
		   tagsetList.add(tagset);
		   
		   ReadPreference regionPref = ReadPreference.nearest(tagsetList);
		   MongoCollection<Document> specificRegionCollection = database.getCollection("<collname>").withReadPreference(regionPref);
		   foundDocument = specificRegionCollection.find().into(new ArrayList<Document>());
		   System.out.println("Documents from specific region :");
		   for(Iterator<Document> idoc = foundDocument.iterator(); idoc.hasNext();)
		   {
			    Document item = idoc.next();
			    System.out.println(item);
		   }
		   
		   
		   System.out.println("Sample complete");
		   
	   } 
	   catch (Exception e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
   }
}
